### PR TITLE
plan/statistics: implement pseudo table and API to estimate row count.

### PR DESCRIPTION
**Pull request by coocood**

### CODE REVIEW

**Comment by coocood** [comment]
@shenli @hanfei19910905 @zimulala PTAL

File: plan/statistics/statistics.go [hunk 1]
```diff
@@ -57,6 +62,86 @@ func (c *Column) String() string {
...
+		prevNumber = c.Numbers[index-1]
+	}
+	lessThanBucketValueCount := number - c.Repeats[index]
+	if match {
+		return lessThanBucketValueCount, nil
+	}
+	return (prevNumber + lessThanBucketValueCount) / 2, nil
+}
+
+// BetweenCount estimate the row count where column greater or equal to a and less than b.
```
Comment (shenli) [inline plan/statistics/statistics.go]: estimates

**Comment by hanfei19910905** [comment]
It's better to mock column interface to behave like filter rate.

**Comment by coocood** [comment]
@hanfei19910905 PTAL

**Comment by hanfei19910905** [comment]
LGTM

File: plan/statistics/statistics.go [hunk 2]
```diff
@@ -25,6 +26,16 @@ import (
 	"github.com/pingcap/tidb/util/types"
 )
 
+const (
+	defaultBucketCount = 256
+
+	pseudoRowCount    = 10000
```
Comment (shenli) [inline plan/statistics/statistics.go]: Add comment for those consts.

File: plan/statistics/statistics.go [hunk 3]
```diff
@@ -57,13 +68,101 @@ func (c *Column) String() string {
...
+		prevNumber = c.Numbers[index-1]
+	}
+	lessThanBucketValueCount := number - c.Repeats[index]
+	if match {
+		return lessThanBucketValueCount, nil
+	}
+	return (prevNumber + lessThanBucketValueCount) / 2, nil
+}
+
+// BetweenRowCount estimate the row count where column greater or equal to a and less than b.
```
Comment (shenli) [inline plan/statistics/statistics.go]: estimates

File: plan/statistics/statistics.go [hunk 4]
```diff
@@ -57,13 +68,101 @@ func (c *Column) String() string {
...
+
 // Table represents statistics for a table.
 type Table struct {
-	info        *model.TableInfo
-	TS          int64 // build timestamp.
-	Columns     []*Column
-	Count       int64 // Total row count in a table.
-	BucketCount int64 // Number of histogram bucket.
+	info    *model.TableInfo
+	TS      int64 // build timestamp.
```
Comment (shenli) [inline plan/statistics/statistics.go]: Why make Ts/Columns/Count public?

Comment (coocood) [inline plan/statistics/statistics.go]: They may be used in other packages, make it public so they can be used anyway user want to.

**Comment by shenli** [comment]
LGTM

**Pull request merged**